scylla/src/transport/topology.rs

```diff
@@ -290,12 +334,213 @@ async fn query_keyspaces(conn: &Connection) -> Result<HashMap<String, Keyspace>,
+        })?;
+
+        let columns = query_columns(conn, keyspace_name.clone(), table_name.clone()).await?;
+
+        result.insert(table_name, Table { columns });
     }
 
     Ok(result)
 }
 
+async fn query_columns(
+    conn: &Connection,
+    keyspace_name: String,
+    table_name: String,
+) -> Result<HashMap<String, Column>, QueryError> {
+    let mut columns_query =
+        Query::new("select column_name, type from system_schema.columns where keyspace_name = ? and table_name = ?");
+    columns_query.set_page_size(1024);
+
+    let rows = conn
+        .query_all(&columns_query, (keyspace_name.clone(), table_name.clone()))
+        .await?
+        .rows
+        .ok_or(QueryError::ProtocolError(
+            "system_schema.columns query response was not Rows",
+        ))?;
+
+    let mut result = HashMap::with_capacity(rows.len());
+
+    for row in rows.into_typed::<(String, String)>() {
+        let (column_name, type_) = row.map_err(|_| {
+            QueryError::ProtocolError("system_schema.columns has invalid column type")
+        })?;
+
+        let column_type = map_string_to_column(conn, keyspace_name.clone(), type_).await?;
+        result.insert(column_name, Column { type_: column_type });
+    }
+
+    Ok(result)
+}
+
+#[async_recursion]
+async fn map_string_to_column(
+    conn: &Connection,
+    keyspace_name: String,
+    type_: String,
+) -> Result<ColumnType, QueryError> {
+    use CqlType::*;
+    let mut frozen = false;
+    let type_ = match type_.as_str() {
+        "ascii" => Ascii,
+        "bigint" => BigInt,
+        "blob" => Blob,
+        "boolean" => Boolean,
+        "counter" => Counter,
+        "date" => Date,
+        "decimal" => Decimal,
+        "double" => Double,
+        "duration" => Duration,
+        "float" => Float,
+        "inet" => Inet,
+        "int" => Int,
+        "smallint" => SmallInt,
+        "text" => Text,
+        "time" => Time,
+        "timeuuid" => Timeuuid,
+        "timestamp" => Timestamp,
+        "tinyint" => TinyInt,
+        "uuid" => Uuid,
+        "varchar" => Text,
+        "varint" => Varint,
+        type_ if type_.ends_with('>') && type_.contains('<') => {
+            let langle_position = type_.find('<').unwrap();
+            let inner_type = &type_[langle_position + 1..type_.len() - 1];
+            let outer_type = &type_[..langle_position];
+            match outer_type {
+                "frozen" => {
+                    frozen = true;
+                    map_string_to_column(conn, keyspace_name, inner_type.to_string())
+                        .await?
+                        .cql_type
+                }
+                "map" | "tuple" => {
+                    let types = inner_type.split(", ").collect::<Vec<_>>();
+                    if outer_type == "map" {
+                        if types.len() != 2 {
+                            return Err(QueryError::InvalidMessage(format!(
```

Comment:
I think that `ProtocolError` would be more suitable instead of your uses of `InvalidMessage`. AFAIK `InvalidMessage` means that an unparsable frame was received from the node, while `ProtocolError` means that the frame was parsed correctly, but something about its contents ws unexpected.

Reply:
The thing with `ProtocolError` is that it only accepts `&'static str`...

Reply:
I see... Changing all ProtocolErrors to accept String would be a lot of effort, so let's keep using InvalidMessage there. I doubt that users will ever want to differentiate between those variants, anyway.

Please add FIXME comments to the places where you return InvalidMessage telling that ProtocolError should be returned there but cannot due to this issue.